solr/core/src/java/org/apache/solr/core/CancellableQueryTracker.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.core;
+
+import org.apache.solr.client.solrj.util.Cancellable;
+import org.apache.solr.search.CancellableCollector;
+import org.apache.solr.request.SolrQueryRequest;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.apache.solr.common.params.CommonParams.QUERY_UUID;
+
+/**
+ * Tracks metadata for active queries and provides methods for access
+ */
+public class CancellableQueryTracker {
+    //TODO: This needs to become a time aware storage model
+    private final Map<String, Cancellable> activeCancellableQueries = new ConcurrentHashMap<>();
+    private final Map<String, String> activeQueriesGenerated = new ConcurrentHashMap<>();
+
+    /** Generates a UUID for the given query or if the user provided a UUID
+     * for this query, uses that.
+     */
+    public String generateQueryID(SolrQueryRequest req) {
+        String queryID;
+        String customQueryUUID = req.getParams().get(QUERY_UUID, null);
+
+        if (customQueryUUID != null) {
+            queryID = customQueryUUID;
+        } else {
+            queryID = UUID.randomUUID().toString();
```

Comment:
I mentioned this previously before the file moved - UUID.randomUUID is eventually slow because it eats system entropy. And it's also synchronized under the hood, getting the bytes, so not great for multi-threaded use. We can leave it in for now and replace with a different implementation later.

Reply:
Ack, will add a todo